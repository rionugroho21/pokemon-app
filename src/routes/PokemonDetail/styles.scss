.pokemon{
  &-detail {
    position: relative;
    width: 100%;
    max-width: 1020px;
    margin: 0px auto;
    display: block;
    padding: 10px 0px 60px 0px;
  }

  &-name {
    font-size: 22px;
    text-align: center;
    width: 100%;
    padding: 10px 0px;
    border-radius: 3px;
    background-color: #ebf5f0;
    text-transform: uppercase;
    cursor: pointer;

    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);

    &:hover {
      box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
      text-decoration: none;
    }
  }

  &-image{
    width: 100%;
    padding: 15px 0px;

    &-item {
      width: 22%;
      display: inline-block;
      vertical-align: top;
      margin-right: 4%;
      text-align: center;
      padding: 0px 0px 10px 0px;
      cursor: pointer;
      box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
      transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
      border-radius: 5px;

      &:hover {
        box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
        text-decoration: none;
      }

      &:last-child {
        margin-right: 0;
      }

      &-img {
        width: 100%;
        max-width: 200px;
        margin: 0px auto;
      }

      img {
        width: 120px;
        height: auto;
        margin: 0px auto;
      }

      span {
        text-align: center;
      }
    }
  }

  &-info {
    cursor: pointer;
    margin-top: 10px;
    margin-bottom: 30px;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
    border-radius: 5px;

    &:hover {
      box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
      text-decoration: none;
    }

    &-title {
      font-size: 18px;
      text-align: center;
      width: 100%;
      padding: 10px 0px;
      border-radius: 5px;
      background-color: #e0e0e0;
      border-bottom-left-radius: 0px;
      border-bottom-right-radius: 0px;
      border-bottom: 0px;
    }

    &-list {
      list-style: none;
      background: #ffffff;
      border: 1px solid #ececec;
      border-radius: 5px;
      border-top-left-radius: 0px;
      border-top-right-radius: 0px;
      padding: 0px 10px;

      li {
        display: inline-block;
        padding: 10px 0px;
        margin-bottom: 10px;
      }
    }

    &-type {
      width: 50%;
      text-align: center;
    }

    &-moves {
      width: 25%;
      text-align: center;
    }

    &-abilility {
      width: 50%;
      text-align: center;
    }

    &-stat {
      width: 100%;
      text-align: left;
    }

    &-progress {
      width: 100%;
      display: inline-block;
      vertical-align: middle;
      margin-top: 5px;
    }
  }

  &-catch {
    position: fixed;
    bottom: 0px;
    width: 1020px;
    padding: 10px 0px;
    background-color: #ef5350;
    border: 0px;
    border-radius: 5px;
    font-size: 22px;
    color: #fff;
    cursor: pointer;
  }
}

@media only screen and (max-width: 1024px) {
  .pokemon {
    &-detail {
      position: relative;
      width: 100%;
      max-width: 100%;
      padding: 10px 16px 60px 16px;
    }

    &-image{
      &-item {
        width: 48%;
        margin-right: 4%;
        margin-bottom: 15px;
  
        &:nth-child(2n) {
          margin-right: 0;
        }
      }
    }

    &-catch {
      width: 100%;
      left: 0px;
      right: 0px;
      border-bottom-left-radius: 0px;
      border-bottom-right-radius: 0px;
    }
  }
}

@media only screen and (max-width: 640px) {
  .pokemon {
    &-info {
      &-moves {
        width: 48%;
      }
    }
  }
}