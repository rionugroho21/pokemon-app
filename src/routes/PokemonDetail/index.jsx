import React, { useContext } from 'react';
import Modal from 'react-bootstrap/Modal';
import ProgressBar from 'react-bootstrap/ProgressBar';
import PopupFailed from './components/PopupFailed/PopupFailed';
import PopupCatch from './components/PopupCatch/PopupCatch';

import get from 'lodash/get';

import usePokemonDetail from './hooks/usePokemonDetail';

import { GlobalContext } from '../../context/GlobalState';

import './styles.scss';

export const PokemonDetail = props => {
  const {
    loading,
    pokemonDetail,
    open,
    isCaught,
    message,
    isExist,
    isNullName,
    handleCatch,
    handleModal,
    handleSave,
    handleNaming
  } = usePokemonDetail(props);

  const abilility = get(pokemonDetail, 'abilities', []);
  const types = get(pokemonDetail, 'types', []);
  const stats = get(pokemonDetail, 'stats', []);
  const moves = get(pokemonDetail, 'moves', []);

  return (
    <>
      {loading ? (
        <p className="loading">Loading...</p>
      ) : (
        <div className="pokemon-detail">
          <div className="pokemon-name">{get(pokemonDetail, 'name', '')}</div>

          <div className="pokemon-image">
            <div className="pokemon-image-item">
              <div className="pokemon-image-item-img">
                <img src={get(pokemonDetail, 'sprites.front_default', '')} alt="Pokemon Normal Front" />
              </div>
              <span>Normal Front</span>
            </div>
            <div className="pokemon-image-item">
              <div className="pokemon-image-item-img">
                <img src={get(pokemonDetail, 'sprites.back_default', '')} alt="Pokemon Normal Back" />
              </div>
              <span>Normal Back</span>
            </div>
            <div className="pokemon-image-item">
              <div className="pokemon-image-item-img">
                <img src={get(pokemonDetail, 'sprites.front_shiny', '')} alt="Pokemon Shiny Front" />
              </div>
              <span>Shiny Front</span>
            </div>
            <div className="pokemon-image-item">
              <div className="pokemon-image-item-img">
                <img src={get(pokemonDetail, 'sprites.back_shiny', '')} alt="Pokemon Shiny Back" />
              </div>
              <span>Shiny Back</span>
            </div>
          </div>

          <div className="pokemon-info">
            <div className="pokemon-info-title">Types</div>
            <ul className="pokemon-info-list">
              {types.map((data, key) => (
                <li key={key} className="pokemon-info-type">{data.type.name.toUpperCase()}</li>
              ))}
            </ul>
          </div>

          <div className="pokemon-info">
            <div className="pokemon-info-title">Moves</div>  
            <ul className="pokemon-info-list">
              {moves.map((data, key) => (
                <li key={key} className="pokemon-info-moves">
                  {data.move.name.toUpperCase()}
                </li>
              ))}
            </ul>
          </div>

          <div className="pokemon-info">
            <div className="pokemon-info-title">Abilility</div>  
            <ul className="pokemon-info-list">
              {abilility.map((data, key) => (
                <li key={key} className="pokemon-info-abilility">{data.ability.name.toUpperCase()}</li>
              ))}
            </ul>
          </div>

          <div className="pokemon-info">
            <div className="pokemon-info-title">Stats</div>  
            <ul className="pokemon-info-list">
              {stats.map((data, key) => (
                <li key={key} className="pokemon-info-stat">
                  {data.stat.name.toUpperCase()} 
                  <div className="pokemon-info-progress">
                    <ProgressBar animated now={data.base_stat} label={`${data.base_stat}%`}/>
                  </div>
                </li>
              ))}
            </ul>
          </div>

          <button className="pokemon-catch" onClick={() => handleCatch()}>Catch the Pokemon</button>

          <PopupFailed 
            open={open} 
            message={message} 
            handleModal={handleModal} 
          />
          <PopupCatch 
            isCaught={isCaught} 
            message={message} 
            pokemonDetail={pokemonDetail}
            isExist={isExist}
            isNullName={isNullName}
            handleModal={handleModal} 
            handleSave={handleSave}
            handleNaming={handleNaming}
          />
        </div>
      )}
    </>
  );
};

export default PokemonDetail;