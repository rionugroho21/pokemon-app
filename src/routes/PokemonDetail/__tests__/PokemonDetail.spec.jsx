import React from 'react';
import { render, cleanup } from '@testing-library/react';

import PokemonDetail from '../index';

const props = {
  loading: false,
  pokemonDetail: {},
  open: false,
  isCaught: false,
  message: 'Failed to catch the Pokemon',
  isExist: false,
  handleCatch: jest.fn(),
  handleModal: jest.fn(),
  handleSave: jest.fn(),
  handleNaming: jest.fn(),
  match: {
    params: {
      id: '1'
    }
  }
};

describe('PokemonDetail Page', () => {
  afterEach(cleanup);

  it('should render page correctly', () => {
    const { asFragment } = render(<PokemonDetail {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});