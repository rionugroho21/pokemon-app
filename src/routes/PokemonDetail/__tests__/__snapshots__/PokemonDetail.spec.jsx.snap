// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonDetail Page should render page correctly 1`] = `
<DocumentFragment>
  <p
    class="loading"
  >
    Loading...
  </p>
</DocumentFragment>
`;
